### PR TITLE
Relax dcou dep. in order-crates-for-publishing.py

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

Recently, `order-crates-for-publishing.py` bite us (or me, specifically, lol) for rejecting a valid circular dep which has been legalized at https://github.com/rust-lang/cargo/pull/7333 (at rust 1.40)

as for the underlying problem, unfortunately, there's a years-long extremely muddying affair regarding circular dev-dep. and lack of grouped (i.e. `workspace`-d) `cargo publish` , our tendency of bench code placement (under `./benches`) and `#[cfg(test)]` interaction, never-stabilized `benches`, no official equivalent to `#[cfg(feature = "dev-context-only-utils")]`, and `cargo hack`-based publish and its drawbacks..... YES, I'm tortuning myself for high standards of our coding practice. ;)

I don't spend writing all the findings. for bored, you can start your rabbit hole starting at these links:
https://github.com/rust-lang/cargo/issues/4242
https://github.com/rust-lang/cargo/pull/7333
https://github.com/rust-lang/rust/issues/84629
https://github.com/rust-lang/rust/issues/67295
https://github.com/rust-lang/crates.io/issues/1789
https://github.com/getsentry/sentry-rust/issues/374
https://users.rust-lang.org/t/conditional-compilation-for-benchmarks/47227/2
https://users.rust-lang.org/t/what-are-the-rules-for-cfg-test/54122/2
https://stackoverflow.com/questions/57607182/equivalent-of-cfgtest-for-benchmarks

tldr: addressing this correctly isn't viable. there's a hope for quite acceptable yet another work-around, if you ignore all the unsatisfying realities. :)

This actually `cargo publish`-safe dep pattern is needed as one of `dev-context-only-utils` usage, specifically blocking https://github.com/solana-labs/solana/pull/32428.

#### Summary of Changes

Actually, we just need to relax `order-crates-for-publishing.py` in rather very conditioned basis (to reduce any risks in `cargo publish`-ing). in a say, the script was too strict, causing false-positive errors.

### CODE REVIEW

File: ci/order-crates-for-publishing.py [hunk 1]
```diff
@@ -21,6 +21,36 @@ def load_metadata():
     return json.loads(subprocess.Popen(
         cmd, shell=True, stdout=subprocess.PIPE).communicate()[0])
 
+no_explicit_version_specified = '*'
+
+def is_special_cased_self_dep(package, dependency):
+    is_special_cased = False
+
+    # Sometimes self-dev-dep with dev-context-only-utils is needed and this dep
```
Comment (mvines) [inline ci/order-crates-for-publishing.py:29]: The use of "self-dev-dep" in this comment throws me off. Maybe it can be expanded or more context added.

It looks like ` is_special_cased_self_dep` only cares about the `dev-context-only-util` crate, so I'd rather it be more explicit in the function name that this is really only about that one crate

**Review by t-nelson — COMMENTED** [review]
can this be generalized to only check for circular dependencies on the same dependency kind? if i'm reading [the docs](https://doc.rust-lang.org/cargo/reference/resolver.html#dev-dependency-cycles) correctly, that's fine

aside: might fuck around and port this one to bash + jq, too :wink:

File: ci/order-crates-for-publishing.py (hunk 1 above)
Comment (ryoqun) [reply ci/order-crates-for-publishing.py]: haha, no wonder for throwing you off. the proper commenting resulted in quite a high wall of text: 069cefafd6ca5ffb56c142d4ae88c6d0f8754257

yeah, i needed fresh eyes to spell the hidden context around `dev-context-only-utils` out. thanks for review. :)

**Comment by ryoqun** [comment]
> can this be generalized to only check for circular dependencies on the same dependency kind? if i'm reading [the docs](https://doc.rust-lang.org/cargo/reference/resolver.html#dev-dependency-cycles) correctly, that's fine

yeah, we can. fyi, the new code in this pr is restricting to the `dev` kind.

However, i wonder the merit of the extra hassle towards generalization... as a matter of fact, seems no one complained for years.

also, note that this script is best-effort, namely it can only detect A <=> B; it can't for A => B => C => A.... I'd like to blind here. ;)

> aside: might fuck around and port this one to bash + jq, too wink

heh, you hate python? no objection for the port for fun. but i want to ship this firstly to unblock #32428 and #31239....

File: ci/order-crates-for-publishing.py [hunk 2]
```diff
@@ -21,6 +21,59 @@ def load_metadata():
...
+        if dependency['req'] == no_explicit_version:
+            is_special_cased = True
+        else:
+            # it's likely `{ workspace = true, ... }` is used, which implicitly pulls the
+            # version in...
+            sys.exit(
+                'Error: wrong dev-context-only-utils circular dependency. try: ' +
+                    '{} = {{ path = ".", features = {} }}\n'
+                    .format(dependency['name'], json.dumps(dependency['features']))
+            )
```
Comment (t-nelson) [inline ci/order-crates-for-publishing.py:67]: can we collect the crates that fail this check and print them at the end instead of early-returning?  spare someone the demoralizing check-fail-fix cycles :wink:

Comment (ryoqun) [reply ci/order-crates-for-publishing.py]: phew, you rightfully spotted my laziness: https://github.com/solana-labs/solana/pull/32432/commits/44e65132b2ce582d2e90ca11ff107921ef80a616

**Review by t-nelson — APPROVED** [review]
this'll do while we still allow python in the monorepo :wink:
